Feat/better dev err

danielphilipjohnson opened the pull request:
### 📚 Description

- Added Youch v4.1.0-beta.4 as a dependency for enhanced error pages

- Port error handling improvements from Pooya's PR, moving src/runtime/internal/error/dev.ts to packages/nuxt/src/core/runtime/nitro/error.ts in the nuxt/nuxt codebase.

### Possible enhancements

![Screenshot from 2025-01-15 13-58-40](https://github.com/user-attachments/assets/d000a78c-7f16-4be9-a52e-28fee047e6f2)

Current nuxt error page.

![Screenshot from 2025-01-15 13-58-11](https://github.com/user-attachments/assets/910323a2-8774-47b6-8c88-1f49c6d2518e)

Current youch error page.

Current Nuxt error page shows the status code prominently at the top. We could maintain this in the Youch implementation by setting:

```js

title: ${errorObject.statusCode}
```

## Dependencies

- Added youch@4.1.0-beta.4

danielphilipjohnson committed:
feat: add youch@4.1.0-beta.4 package for error handling

danielphilipjohnson committed:
feat: update error handler to use Youch for HTML formatting

HigherOrderLogic commented on packages/nuxt/src/core/runtime/nitro/error.ts:
```diff
@@ -74,15 +80,22 @@ export default <NitroErrorHandler> async function errorhandler (error: H3Error,
   }
 
-  const html = await res.text()
+  const html = await youch.toHTML(error, {
+    request: {
+      url,
+      method: event.method,
+      headers: getRequestHeaders(event),
+    },
+  })
+
   if (event.handled) { return }
 
```
```suggestion
  if (event.handled) { return }
  
  const html = await youch.toHTML(error, {
    request: {
      url,
      method: event.method,
      headers: getRequestHeaders(event),
    },
  })
```

danielphilipjohnson replied:
Sorry for the lateness will push this change shortly :pray:

danielphilipjohnson committed:
fix(error): reorder error handling logic to respect handled flag

Move HTML generation after event.handled check to prevent unnecessary
error page generation for already handled errors.

danielroe commented:
thank you ❤️ 

linking also https://github.com/nuxt/nuxt/pull/30624 - and see comment in https://github.com/nuxt/nuxt/pull/30624#issuecomment-2594995699

danielroe commented:
marking as draft until implementation in nitro is finalised. ❤️

danielroe commented:
resolved in https://github.com/nuxt/nuxt/pull/31230
